### DISCUSSION TITLE
Allow integrations to register their own config panels

### DISCUSSION THREAD

**Discussion by legrego** [home-assistant/frontend#7820] [Configuration]
Integrations are capable of exposing their config options by registering an "Options Flow Handler", as described in the [docs](https://developers.home-assistant.io/docs/config_entries_options_flow_handler).

While most integrations follow this convention, there are a few hard-coded integrations which are able to [bypass this, in favor of rendering their own custom panels](https://github.com/home-assistant/frontend/blob/6f0c79ec25de0b84e1e74b1685290965c7616dd8/src/panels/config/integrations/ha-integration-card.ts#L52-L69). This allows them to provide a richer configuration experience, which is essential for integrations with more complex setups.

I am in the process of updating my custom component for Elasticsearch to take advantage of config flows (https://github.com/legrego/homeassistant-elasticsearch/issues/74), but the amount of customization that I'm able to offer via `configuration.yaml` doesn't seem to scale to the current config flow/options flow capabilities. While I could _technically_ make it work, it would lead to a really poor user experience. Having the ability to write my own panel would go a long way here.

**Comment by bramkragten** [comment]
You can already register a custom panel, the only thing you can't do is add the button to the integration card.

So you should already be able to do this.

**Reply by legrego** [reply]
@bramkragten thanks for the response.

> the only thing you can't do is add the button to the integration card

I saw that I could register a custom panel, but I think the ability to add the button to the integration card would help create a cohesive experience. Without this capability, we'll have certain "special" integrations with the ability to add this button, while all other integrations are left to implement it differently.

If the custom panel + sidebar link is the way we want integrations to handle this, then why are some integrations allowed to break those rules?


For what it's worth, I'm fine with whatever is decided. I just want to make sure I'm designing this in a way that is as consistent as possible, and in a way that is recommended/supported.

**Reply by bramkragten** [reply]
You are talking about a custom integration right? 
Integrations that are not custom are not allowed to just register a configuration panel, they will either have to handle the config with the config flow, are create a dedicated panel like mqtt, zha, ozw.

The problem with these dedicated panels is that they need to be maintained, so we don't want to add them for every single integration.

**Reply by legrego** [reply]
Yes, I'm talking about a custom integration. I'm not advocating for all integrations to have a dedicated panel -- I understand that's not a maintainable or sustainable approach, so having config flows/option flows makes a lot of sense for most integrations.

> Integrations that are not custom are not allowed to just register a configuration panel, they will either have to handle the config with the config flow, are create a dedicated panel like mqtt, zha, ozw

I understand -- I don't want all integrations to have these dedicated panels. Rather, I'm asking for a supported mechanism for custom integrations to add their own dedicated panels, in the same way that mqtt, zha, and ozw have. I know I can create a custom panel which renders in the main sidebar, but that is inconsistent with the way mqtt, ozw, etc behave today.

If mqtt, ozw, etc rendered their panels from links in the main sidebar, then this wouldn't be an issue. My concern is that these specific integrations are allowed to render their custom panels from the integrations screen, whereas you're asking for all other integrations to use a different mechanism altogether. The inconsistency is mainly what bothers me here.

I want integrations to be able to tell home-assistant that they should have a `Configure` button rendered on the integrations screen, and to have that button render their custom panel.
